In [491]:
import nltk
from nltk.corpus import brown
from nltk.corpus import words
from nltk.corpus import webtext
from nltk.corpus import wordnet as wn
from nltk.corpus import cmudict
from nltk.corpus import inaugural as ig
from nltk.book import *


ข้อ 2.

ข้อ 2.1 อยู่ในรูปแบบ Overlapping เพราะข้อมูลมีคำซ้อนกันมาก

ข้อ 2.2 หาโดยการหาคำที่ขึ้นต้นด้วยตัว I ตัวใหญ่ เพราะลักษณะของชื่อเฉพาะมักจะขึ้นด้วยตัวอักษรตัวใหญ่ และบางครั้งอาจจะเป็นตัวใหญ่ทั้งหมด

ข้อ 2.3

In [490]:
list = [w for w in text1.vocab().keys() if str(w).startswith('I')]
fdlist = FreqDist(list)
print('Num : ' + str(len(fdlist.keys())))
fdlist

Num : 95


FreqDist({'I': 1,
          'IAN': 1,
          'IBID': 1,
          'ICELAND': 1,
          'ICELANDIC': 1,
          'II': 1,
          'III': 1,
          'ILE': 1,
          'IMPOSING': 1,
          'IN': 1,
          'INN': 1,
          'INTO': 1,
          'IS': 1,
          'ISAIAH': 1,
          'ISHMAEL': 1,
          'ISLAND': 1,
          'ISLANDS': 1,
          'ISOLATO': 1,
          'ISOLATOES': 1,
          'IT': 1,
          'ITS': 1,
          'IV': 1,
          'IVORY': 1,
          'Iceberg': 1,
          'Icebergs': 1,
          'Iceland': 1,
          'Icelandic': 1,
          'Icy': 1,
          'If': 1,
          'Ifs': 1,
          'Ignorance': 1,
          'Illinois': 1,
          'Immediately': 1,
          'Immemorial': 1,
          'Immense': 1,
          'Immortal': 1,
          'Impenetrable': 1,
          'Imperial': 1,
          'Impiety': 1,
          'Impossible': 1,
          'Imprimis': 1,
          'Improving': 1,
          'In': 1,
          'Inasm

ข้อ 2.4 มีรายชื่อดังนี้ ISAIAH ISHMAEL Ixion

ข้อ 3.

In [11]:
brown.categories()
list = ['adventure', 'fiction', 'news', 'reviews', 'religion', 'romance']

In [13]:
for category in list:
    print(category +' : token = '+ str(len(brown.words(categories=category))) + ' vocab = ' +
        str(len(set(brown.words(categories=category)))) + ' Lexical diversity = ' +
        format(len(set(brown.words(categories=category)))/
                len(brown.words(categories=category)),'.2f'))

adventure : token = 69342 vocab = 8874 Lexical diversity = 0.13


fiction : token = 68488 vocab = 9302 Lexical diversity = 0.14


news : token = 100554 vocab = 14394 Lexical diversity = 0.14


reviews : token = 40704 vocab = 8626 Lexical diversity = 0.21


religion : token = 39399 vocab = 6373 Lexical diversity = 0.16


romance : token = 70022 vocab = 8452 Lexical diversity = 0.12


Category ที่มีความหลากหลายน้อยที่สุดก็คือ Romance

ข้อ 4.

In [30]:
def unknown(text):
    text_vocab = set(w.lower() for w in text if w.isalpha())
    english_vocab = set(w.lower() for w in words.words())
    unusual = text_vocab - english_vocab
    return sorted(unusual)

In [31]:
len(unknown(webtext.words()))

7538

ข้อ 4.1 พบ 7538 คำ

ข้อ 5 

ข้อ 5.1

In [35]:
len_list = []
for word in unknown(webtext.words()):
    len_list.append(len(word))
fdlist = nltk.FreqDist(len_list)

ข้อ 5.2

In [37]:
fdlist.max()

7

คำที่มีความถี่สูงที่สุดมีความยาว 7 ตัวอักษร

ข้อ 5.3 

In [38]:
fdlist.most_common()

[(7, 1292),
 (6, 1135),
 (8, 1025),
 (5, 880),
 (9, 836),
 (4, 528),
 (10, 526),
 (3, 373),
 (11, 334),
 (12, 169),
 (2, 133),
 (13, 119),
 (14, 47),
 (15, 42),
 (16, 28),
 (17, 15),
 (20, 10),
 (1, 9),
 (19, 7),
 (22, 7),
 (18, 5),
 (24, 5),
 (21, 4),
 (23, 3),
 (26, 3),
 (25, 1),
 (27, 1),
 (28, 1)]

คำที่มีความถี่ต่ำสุดมีความยาว 25,27,28 ตัวอักษร

ข้อ 8

In [41]:
[w for w in 'The quick brown fox jumps over the lazy dog'.split() if(len(w) == 3)]


['The', 'fox', 'the', 'dog']

ข้อ 9

In [446]:
def my_bigram(text, word):
    bigram = bigrams(text)
    cfd = nltk.ConditionalFreqDist(bigram)
    list = cfd.conditions()
    list2 = [w for w in list if(cfd[w][word] >= 1)]
    f_list = []
    for w in list2:
        for q in range(cfd[w][word]):
            f_list.append(w)
    fdlist = FreqDist(f_list)
    return fdlist

In [457]:
fdlist = my_bigram(ig.words(), 'citizens')
print(fdlist.most_common()[:5])

[('fellow', 60), ('-', 34), ('our', 21), ('Fellow', 16), ('its', 12)]


ข้อ 9.1 citizens คำว่า fellow ความถี่ 60

In [458]:
fdlist = my_bigram(ig.words(), 'people')
print(fdlist.most_common()[:5])

[('the', 243), ('our', 73), ('American', 39), ('a', 23), ('The', 16)]


ข้อ 9.2 people คำว่า the ความถี่ 243

In [459]:
fdlist = my_bigram(ig.words(), 'government')
print(fdlist.most_common()[:5])

[('of', 65), ('-', 30), ('a', 24), ('the', 15), ('free', 11)]


ข้อ 9.3 government คำว่า of ความถี่ 65

ข้อ 10

In [ ]:
wn.synsets('cetacean')

[Synset('cetacean.n.01'), Synset('cetacean.a.01')]

ข้อ 10.1 

In [82]:
definition = wn.synset('cetacean.n.01').definition()

In [86]:
'definition of word is :  ' + noun

'definition of word is :  large aquatic carnivorous mammal with fin-like forelimbs no hind limbs, including: whales; dolphins; porpoises; narwhals'

ข้อ 10.2

In [87]:
wn.synset('cetacean.n.01').lemma_names()

['cetacean', 'cetacean_mammal', 'blower']

คำที่อยู่ใน synonyms มีทั้งหมด 3 คำ

ข้อ 10.3 

In [92]:
mm = wn.synset('mammal.n.01')
hm = wn.synset('homo.n.01')
dp = wn.synset('dolphin.n.01')

In [96]:
print('mammal and homo = ' + str(mm.path_similarity(hm)))
print('mammal and dolphin = ' + str(mm.path_similarity(dp)))
print('homo and dolphin = ' + str(hm.path_similarity(dp)))


mammal and homo = 0.14285714285714285
mammal and dolphin = 0.1111111111111111
homo and dolphin = 0.07692307692307693


คำที่มีมีความหมายใกล้เคียงมากที่สุดคือ mammal และ homo
คำที่มีความหมายใกล้เคียงน้อยที่สุดคือ homo และ dolphin
ใช้คำสั่ง path_similarity ในการตรวจสอบ โดยมีความหมายคือ ใช้ shortest path ในการเชื่อม hypernym (จะรีเทิร์น -1 ในกรณีที่หา path ไม่เจอ)
แล้วนำมาเปรียบเทียบกับ synset ของตัวเองที่มีค่าเป็น 1

ข้อ 10.4

In [97]:
ctc = wn.synset('cetacean.n.01')
ctc_mm = wn.synset('cetacean_mammal.n.01')
bw = wn.synset('blower.n.01')

In [104]:
ctc.root_hypernyms()

[Synset('entity.n.01')]

ข้อ 11

In [106]:
prondict = cmudict.dict()

def syllable(word):  
    try:
        return len([ph for ph in prondict[word.lower()]])
    except Exception as error:
        return 0;

ข้อ 12

In [113]:
list = [syllable(w.lower()) for w in ig.words()]
fdlist = nltk.FreqDist(list)
fdlist.most_common()

[(1, 63364), (2, 46906), (3, 19092), (0, 14280), (4, 2092), (5, 1)]

In [126]:
list2 = len(ig.words())
syl_value = 0
for syl in fdlist.keys():
    syl_value += fdlist[syl] * syl
syl_per_word = syl_value / list2
format(syl_per_word,'.2f')

'1.53'

ข้อ 12.1 จำนวนพยางค์ที่ยาวที่สุด มีความยาว 5 พยางค์

ข้อ 12.2 จำนวนพยางค์ที่มีความถี่สูงสุด มีความยาว 1 พยางค์

ข้อ 12.3 ค่าเฉลี่ยของจำนวนพยางค์ต่อคำ มีความยาว 1.53 พยางค์

ข้อ 13.

ข้อ 13.1

RE = 206.835 – (1.015 x ASL) – (84.6 x ASW) 

RE = Readability Ease 

ASL = Average Sentence Length (i.e., the number of words divided by the number of sentences) 

ASW = Average number of syllables per word (i.e., the number of syllables divided by the number of words) 

The output, i.e., RE is a number ranging from 0 to 100. The higher the number, the easier the text is to read. 

 Scores between 90.0 and 100.0 are considered easily understandable by an average 5th grader.

 Scores between 60.0 and 70.0 are considered easily understood by 8th and 9th graders

 Scores between 0.0 and 30.0 are considered easily understood by college graduates.

90-100 : Very Easy 

80-89 : Easy
 
70-79 : Fairly Easy 

60-69 : Standard
 
50-59 : Fairly Difficult
 
30-49 : Difficult
 
0-29 : Very Confusing
 
Ref : http://www.readabilityformulas.com/flesch-reading-ease-readability-formula.php

ข้อ 13.2

In [1]:
def readability_ease(asl,asw):
    return 206.835 - (1.015 * asl) - (84.6 * asw)

ข้อ 13.3

In [ ]:
def grade_cal(text):
    sents = text.split('.')
    asl = sum(len(x.split()) for x in sents) / len(sents)